tartufo/scanner.py

```diff
@@ -683,3 +685,51 @@ def chunks(self):
         )
         for blob, file_path in self._iter_diff_index(diff_index):
             yield types.Chunk(blob, file_path, {})
+
+
+class FolderScanner(ScannerBase):
+    """Used to scan a folder."""
+
+    target: str
+
+    def __init__(
+        self,
+        global_options: types.GlobalOptions,
+        target: str,
+    ) -> None:
+        """Used for scanning a folder.
+
+        :param global_options: The options provided to the top-level tartufo command
+        :param target: The local filesystem path to scan
+        """
+        self.target = target
+        super().__init__(global_options)
+
+    @property
+    def chunks(self) -> Generator[types.Chunk, None, None]:
+        """Yield the individual file changes currently staged for commit.
+
+        :rtype: Generator[Chunk, None, None]
+        """
+
+        for blob, file_path in self._iter_folder():
+            yield types.Chunk(blob, file_path, {})
+
+    def _iter_folder(self) -> Generator[Tuple[str, str], None, None]:
+        folder_path = pathlib.Path(self.target)
+        for file_path in folder_path.rglob("**/*"):
+            relative_path = file_path.relative_to(folder_path)
+            if file_path.is_file() and self.should_scan(str(relative_path)):
+                try:
+                    with file_path.open("rb") as fhd:
+                        data = fhd.read()
+                except OSError as exc:
+                    raise click.FileError(filename=str(file_path), hint=str(exc))
```

Comment:
Are we sure that we want to raise an exception here and abort the scan? Or should we (similar to a utility like `df`) just print out an error message and continue on? I could honestly go either way here.

If we are going to go with the route of aborting, we'll want two quick enhancements. First, just a general best practice (possibly not followed everywhere in the codebase at the moment, but should be) would be to use the `raise ... from` syntax:

```suggestion
                    raise click.FileError(filename=str(file_path), hint=str(exc)) from exc
```

The other is that in the command itself, we will want an additional

```python
except click.FileError:
```
Similar to what is done with `scan-remote-repo`, to make sure we exit gracefully instead of bombing out with a stack trace.

Reply:
If the file might have a secret, I don't want it to be hidden by log entry. The user should omit the file if they don't want it scanned. As far as the stack trace, I just copied that from other code. I think just showing the exception message as the hint and log the full exception might be better. I agree, I don't like the stack trace. I also thought click handles `FileError` gracefully.

Reply:
This is what it would look like:
```
Error: Could not open file /Users/dclayton/git/github.com/dclayton-godaddy/tartufo/tests/data/scan_folder/test.txt: [Errno 13] Permission denied: '/Users/dclayton/git/github.com/dclayton-godaddy/tartufo/tests/data/scan_folder/test.txt'
```

Reply:
Ah interesting, there must be some magical internal handling for file errors. That works for me!

Reply:
I agree it's better to abort the scan rather than obfuscate the fact that a selected target might contain secrets that we didn't report. Plus it's simple. FWIW, it seems the "continue on" alternative would require us to dummy up a finding that represents the entire file (if we can't assess it, we assume it's problematic) and that just doesn't seem like it's worth the effort in this context.